mkdocs/livereload/__init__.py

```diff
@@ -86,6 +86,21 @@ def callback(event):
         handler.on_any_event = callback
         self.observer.schedule(handler, path, recursive=recursive)
 
+        def schedule(handler, path, recursive=True):
+            if schedule.depth >= schedule.limit:
+                return
+            schedule.depth += 1
+            if os.path.islink(path):
+                self.observer.schedule(handler, os.path.realpath(path), recursive=recursive)
+            elif recursive and os.path.isdir(path):
+                for name in os.listdir(path):
+                    schedule(handler, os.path.join(path, name), recursive=recursive)
+            schedule.depth -= 1
+
+        schedule.limit = 9
+        schedule.depth = 1
+        schedule(handler, os.path.realpath(path), recursive=recursive)
```

Comment:
Again untested code. This is the real simplification applicable only if we *actually* want to ignore nested symlinks.

```suggestion
        if recursive:
            for root, dirs, _ in os.walk(path):
                for relpath in dirs:
                    subpath = os.path.join(root, relpath)
                    if os.path.islink(subpath):
                        self.observer.schedule(handler, subpath, recursive=recursive)
```

Two additional comments on this front:

* I think traversing anything at all shouldn't be done with `recursive=False`. So I make it as such here.
* What was the reason to limit the depth of directory trees to 9? There is no threat of infinite recursion.

Reply:
On my last point:
Ah no, I understand how traversal works in your code. All is correct there (or at least is more likely to make sense), as traverses 1 file symlink at the root. My code fails to do that.

Reply:
> * I think traversing anything at all shouldn't be done with recursive=False. So I make it as such here.

This might have been flawed, too tired to check, just take care that "traversing" (starting at) the first level must not mean it is recursive.

What I specifically don't know is the effect of recursive in watchdogs observe for a path that is a directory.

The implementation was under my assumption that the non-recursive observing of a directory path results in the observation of all files in that directory plus the addition and removal of files there-in.

You already added an additional remark yourself, I'm still interested in case you want to share your understanding of how the effect of recursive in watchdogs observe comes into play.

> * What was the reason to limit the depth of directory trees to 9? There is no threat of infinite recursion.

About why there is a limit at all: My very first approach had no function at all, so there was no recursion. Then I added the function as I realized that w/ recursion it is very straight forward for directory tree iteration (mind that I'm a python newbie). Then I reviewed the implementation and I realized that there were no limits. So I decided to add one, as even the code could endlessly run, any system it runs on has limits.

About why 9: I just needed a number. Three was too low. Six seemed cheap, seven nice and then gave two more. One for the part that the with any such limit it might be one less on how to start the traversal and one for the goddesses or guests.

Yes, the limit is synthetic, more research and development is likely to its benefit.

Reason: Filesystem I/O. Mitigation of the risk of tree traversal circles = endless loop/recursion. You're right when you look at the code, it does not look prone to it. But the code only shows one half of the story. While it runs, the file-system beneath is not frozen. That is the other half and there you have these race-conditions and what not.

The limit of nine might be wrong by the number, point in case is to have a limit. If its too strict, the number can be raised. That in short perhaps summarizes how I thought about it when/after adding the limiting.

This is missing a bit in the commit message, I shortened it while preparing the patch.

Reply:
> Symlinks to symlinks would've worked before.

I wasn't aware of this before, so from how I understand what a regression is, this just extended the scope. I'm glad this could be figured out that quickly.

> And allowing only 1 level of symlinks just seems arbitrary.

Well, there was one point I didn't explicitly mention before, but when following symlinks more level deeps while traversing a directory structure there is the problem of endless recursion:

~~~
* docs_dir
    * blue
        * red
          * yellow (symbolic link to ../../../docs_dir)  
~~~

this would then require bookkeeping to detect such a circular reference. this is above my paygrade in python.


> In any case, this needs unittests.

Big fan of those, but maybe before adding them, what your insightful comments on the realpath traversal depth and also on the general docs_dir traversal depth have shown to me, I would prefer a regression test-case. Honestly, I don't know what the processes are in this project, so see it as a suggestion on how to proceed and if you like let me know what you think about it.

Reply:
> The implementation was under my assumption that the non-recursive observing of a directory path results in the observation of all files in that directory plus the addition and removal of files there-in.

I think that's right.

---

Regarding loops in file systems.

If we support symbolic links, yes, they are definitely possible and need to be detected.
And then also a recursion-based solution is required; `os.walk` is not usable.

If we don't support, you are also correct that exceptional cases can still cause loops. But I think in that case it is very typical to ignore such a possibility (and certainly way more typical than limiting traversal at N nested directories). I mean, if this was a ubiquitous concern, `os.walk` would basically never be usable. Instead, I'd say it's considered fair game for non-specialized tooling to enter infinite recurstion when encountering a hard loop in the file system; and usually people who have such loops are also aware of the complications. But this is just coming from my observations, I don't have hard evidence.

And if we talk about `livereload` library, well, it just follows all symlinks and all dirs with absolutely no attempt to avoid loops or putting other limits on traversal.
https://github.com/lepture/python-livereload/blob/ecabebaeee2a40743f5a8216e1856c9f9150913f/livereload/watcher.py#L197.

---

Regarding how to actually detect loops in symlinks:
I think this is better done by remembering seen directories so far. Then depth limiting is not needed.
I see you are aware of this idea already, but sure, maybe it needs more specific Python knowledge.

> this is above my paygrade in python

Mhm well I would be glad to take over implementing this. Just didn't want to rush in and be like "no, let *me* do it".
Please let me know what you think.
And I would indeed do it with the approach that doesn't require limiting traversal depth, and I'd make it traverse nested symlinks.

---

> > In any case, this needs unittests.
>
> Big fan of those, but maybe before adding them

Ah I think there's no "before" that applies here. This case to me seems like a poster child for writing tests first (or at least right after you realize that the code might go through several iterations before reaching the final one). You do the setup with whatever directory structures are relevant, define what should be done in each scenario and test accordingly. Then you have a tool to always instantly tell you: "is the code I just wrote correct?". Moreover, it even gives you the ability to answer review comments with something like "no, I can't remove that line, see, then this test fails".

Of course, I realize that we haven't yet decided what *should* indeed be done in each scenario. But for our particular case that's not a significant obstacle, because it's super cheap to adapt the tests if needed.

To me, tests here are almost more valuable than the solution itself. Otherwise we can't even be sure that the issue is fixed.